torch/csrc/autograd/engine.cpp

```diff
@@ -526,6 +519,8 @@ auto Engine::thread_main(const std::shared_ptr<GraphTask>& graph_task) -> void {
         continue;
       }
 
+      set_device(worker_device);
```

Comment:
This function is relatively expensive. I think we should have a version that:
- skip the tls set
- Only creates the Device and device guards if the device is not properly set already. Checking the current device should be a lot cheaper than all of that right?

Reply:
Thanks for the response.
- Skipping tls guard https://github.com/pytorch/pytorch/blob/1e6961586b9d213e05cd5ecbd5a1c17acbcb051d/torch/csrc/autograd/engine.cpp#L533 caused test failures (ran locally). 

- device guard skips setDevice if it is already set: https://github.com/pytorch/pytorch/blob/1e6961586b9d213e05cd5ecbd5a1c17acbcb051d/c10/cuda/impl/CUDAGuardImpl.h#L52-L58
despite that I attempted to skip creating device guard for CUDA in [df53cb8](https://github.com/pytorch/pytorch/pull/92354/commits/df53cb8e0a19e20c6685c2abe111b68be8323dec)